File: 0815-bus-routes/0815-bus-routes.py
class Solution:
    def numBusesToDestination(self, routes: List[List[int]], source: int, target: int) -> int:
        def have_common_node(route1: List[int], route2: List[int]) -> bool:
            i, j = 0, 0
            while i < len(route1) and j < len(route2):
                if route1[i] == route2[j]:
                    return True
            
                if route1[i] < route2[j]:
                    i += 1
                else:
                    j += 1
            return False
        
        # Return true if the given stop is present in the route, False otherwise
        def is_stop_exist(route: List[int], stop: int) -> bool:
            for j in range(len(route)):
                if route[j] == stop:
                    return True
            return False
    
        
        if source == target:
            return 0
        
        for i in range(len(routes)):
            routes[i].sort()
        
        adj_list = [[] for _ in range(501)]
        for i in range(len(routes)):
            for j in range(i + 1, len(routes)):
                if have_common_node(routes[i], routes[j]):
                    adj_list[i].append(j)
                    adj_list[j].append(i)
        
        queue = deque()
        for i in range(len(routes)):
            if is_stop_exist(routes[i], source):
                queue.append(i)
        
        visited = [False] * len(routes)
        bus_count = 1
        
        while queue:
            size = len(queue)
            
            while size > 0:
                node = queue.popleft()
                
                # Return bus_count, if the stop target is present in the current route
                if is_stop_exist(routes[node], target):
                    return bus_count
            
                # Add the adjacent routes
                for adj in adj_list[node]:
                    if not visited[adj]:
                        visited[adj] = True
                        queue.append(adj)
                size -= 1
                
            bus_count += 1
        return -1
        

File: 0815-bus-routes/0815-bus-routes.cpp
class Solution {
private:
    vector<int> adjList[501];
    
    // Iterate over each pair of routes and add an edge between thm if there's a common stop
    void createGraph(vector<vector<int>>& routes){
        for (int i = 0; i < routes.size(); ++i){
            for (int j = i + 1; j < routes.size(); ++j){
                if (haveCommonNode(routes[i], routes[j])){
                    adjList[i].push_back(j);
                    adjList[j].push_back(i);
                }
            }
        }
    }
    
    // Returns true if the provided routes have a common stop, false otherwise.
    bool haveCommonNode(vector<int>& route1, vector<int>& route2){
        int i = 0, j = 0;
        while (i < route1.size() && j < route2.size()){
            if (route1[i] == route2[j])
                return true;
            route1[i] < route2[j]? ++i : ++j;
        }
        return false;
    }
    
    // Add all the routes in the queue having the source as one of the stops
    void addStartingNodes(queue<int>& q, vector<vector<int>>& routes, int source){
        for (int i = 0; i < routes.size(); ++i){
            if (isStopExist(routes[i], source))
                q.push(i);
        }
    }
    
    // Returns True if the given stop is present in the route, false otherwise
    bool isStopExist(vector<int>& route, int stop){
        for (int j = 0; j < route.size(); ++j){
            if (route[j] == stop)
                return true;
        }
        
        return false;
    }
    
public:
    int numBusesToDestination(vector<vector<int>>& routes, int source, int target) {
        if (source == target)
            return 0;
        
        for (int i = 0; i < routes.size(); ++i)
            sort(routes[i].begin(), routes[i].end());
        
        createGraph(routes);
        queue<int> q;
        addStartingNodes(q, routes, source);
        
        vector<bool> vis(routes.size());
        int busCount = 1;
        while(!q.empty()){
            int sz = q.size();
            
            while (sz--){
                int node = q.front();
                q.pop();
                
                // Return busCount, if the stop target is present in the current route.
                if (isStopExist(routes[node], target))
                    return busCount;
                
                for (int adj : adjList[node]){
                    if (!vis[adj]){
                        vis[adj] = true;
                        q.push(adj);
                    }
                }
            }
            
            ++busCount;
        }
        
        return -1;
    }
};